pkg/metricsadapter/provider/resourcemetrics.go

```diff
@@ -0,0 +1,462 @@
+}
+
+// queryNodeMetricsBySelector queries metrics by node selector from target clusters
+func (r *ResourceMetricsProvider) queryNodeMetricsBySelector(selector string) ([]metrics.NodeMetrics, error) {
+	labelSelector, err := labels.Parse(selector)
+	if err != nil {
+		klog.Errorf("Failed to parse label selector: %v", err)
+		return nil, err
+	}
+
+	resourceQueryFunc := func(sci genericmanager.SingleClusterInformerManager, clusterName string) error {
+		nodes, err := sci.Lister(NodesGVR).List(labelSelector)
+		if err != nil {
+			klog.Errorf("Failed to list pods in cluster(%s): %v", clusterName, err)
+			return err
+		}
+		if len(nodes) == 0 {
+			return errors.NewNotFound(PodsGVR.GroupResource(), "")
+		}
+		return nil
+	}
+	metricsQueryFunc := func(sci genericmanager.SingleClusterInformerManager, _ string) (interface{}, error) {
+		metrics, err := sci.GetClient().Resource(nodeMetricsGVR).List(context.Background(), metav1.ListOptions{
+			LabelSelector: selector,
+		})
+		return metrics, err
+	}
+
+	metricsQuery, err := r.getMetricsParallel(resourceQueryFunc, metricsQueryFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	var nodeMetrics []metrics.NodeMetrics
+	for index := range metricsQuery {
+		metricsData := metricsQuery[index].(*unstructured.UnstructuredList)
+		internalMetrics, err := metricsConvertV1beta1NodeToInternalNode(metricsData.Items...)
+		if err != nil {
+			continue
+		}
+		nodeMetrics = append(nodeMetrics, internalMetrics...)
+	}
+
+	return nodeMetrics, nil
+}
+
+// GetPodMetrics queries metrics by the internal constructed pod
+func (r *ResourceMetricsProvider) GetPodMetrics(pods ...*metav1.PartialObjectMetadata) ([]metrics.PodMetrics, error) {
+	var podMetrics []metrics.PodMetrics
+
+	// In the previous step, we construct pods list with only one element.
+	if len(pods) != 1 {
+		return podMetrics, nil
+	}
+
+	// In the previous step, if query with label selector, the name will be set to empty
+	if pods[0].Name == "" {
+		namespace := pods[0].Namespace
+		selectorStr := pods[0].Annotations[labelSelectorAnnotationInternal]
+		return r.queryPodMetricsBySelector(selectorStr, namespace)
+	}
+
+	return r.queryPodMetricsByName(pods[0].Name, pods[0].Namespace)
+}
+
+// GetNodeMetrics queries metrics by the internal constructed node
+func (r *ResourceMetricsProvider) GetNodeMetrics(nodes ...*corev1.Node) ([]metrics.NodeMetrics, error) {
+	var nodeMetrics []metrics.NodeMetrics
+
+	// In the previous step, we construct node list with only one element, this should never happen
+	if len(nodes) != 1 {
+		// never reach here
+		return nodeMetrics, nil
+	}
+
+	// In the previous step, if query with label selector, the name will be set to empty
+	if nodes[0].Name == "" {
+		selectorStr := nodes[0].Annotations[labelSelectorAnnotationInternal]
+		return r.queryNodeMetricsBySelector(selectorStr)
+	}
+
+	return r.queryNodeMetricsByName(nodes[0].Name)
+}
+
+// PodLister is an internal lister for pods
+type PodLister struct {
```

Comment:
This struct implement `cache.GenericLister` interface. I'd like to add the following to ensure this interface is fully implemented.
```
var _ cache.GenericLister = &PodLister{}
var _ v1listers.NodeLister = &NodeLister{}
var _ api.MetricsGetter = &ResourceMetricsProvider{}
```

Reply:
When init the metrics server, we use these two listers, there will be type checking. So whether is it necessary?